### ISSUE TITLE
Inconsistency with JSON.generate in context of a Rails app

### ISSUE THREAD

**Issue by lucaong** [ohler55/oj#651]
In a freshly generated Rails app (created by just running `rails new my_app`), with only the addition of the `oj` gem to the Gemfile, assuming that we have a class like:

```ruby
class Foo
  def initialize
    @x = 123
  end
end
```

Then, `Oj.mimic_JSON()` (or, equivalently, `Oj.optimize_rails()`) fails to mimic the serialization behavior. In a `rails console`:

```ruby
foo = Foo.new

JSON.generate(foo)
# => "\"#<Foo:0x00007fc9e0f294f8>\""

Oj.mimic_JSON()
JSON.generate(foo)
# => "{\"x\":123}"
```

As seen here, Oj serializes the object by serializing its instance variables (probably due to calling `foo.as_json` internally), differently from JSON which just prints the object as a string.

Note that in Rails using the default JSON, the behavior of `object.to_json` and `JSON.generate(object)` differs in that only `object.to_json` relies on calling `object.as_json`, but with `Oj.mimic_JSON` this difference is lost:

```ruby
class X
  def as_json(arg)
    { abc: 123 }
  end
end

x = X.new

# In a Rails app, before requiring Oj:

x.to_json
# => "{\"abc\":123}"

JSON.generate(x)
# => "\"#<X:0x00007fc8a52ca4c8>\""

# When requiring Oj and mimicking JSON:

require 'oj'
Oj.mimic_JSON()

x.to_json
# => "{\"abc\":123}"

JSON.generate(x)
# => "{\"abc\":123}"
```

In other words, to correctly mimic the behavior of the JSON gem in a Rails app, Oj should call `object.as_json` when calling `object.to_json`, but should NOT call `object.as_json` when calling `JSON.generate(object)`.

I am running this code in Ruby 2.6.6 and Rails 6.1.3.1, not sure if results vary by version. In a plain Ruby project (not requiring Rails) this problem does not seem to occur, and `JSON.generate` behaves like the JSON gem version when serializing instances of custom classes. Conversely, if `require 'rails'` is added before calling `Oj.mimic_JSON()`, the issue occurs.

**Comment by ohler55** [comment]
I'll dig into the issue this weekend.

**Comment by ohler55** [comment]
I can verify that without Rails all seems fine. It also appears as if in requiring rails does not change the JSON.generate so that is indeed an issue. I'll continue working on a fix.

**Comment by lucaong** [comment]
Thanks for investigating this issue, and for `oj`!

**Comment by ohler55** [comment]
I'm not sure if it is rails or the json gem but what appears to be happening is that `JSON.generate` after pulling in rails will ignore the `to_json` monkey patched by ActiveSupport but will honour all others. I'm sure there is some code somewhere that defines something more clear but I have not been able to find it. I think the best I can do is in `JSON.generate` just ignore any `to_json` defined by ActiveSupport. Special case always slow the code down but that's the penalty for trying to be compatible I suppose.

**Comment by ohler55** [comment]
It's pretty ugly but a fix is in the generate-special branch. Let me know if you have any issues with it. If not I will cut a release.

**Comment by lucaong** [comment]
Oh wow yeah the behavior is more complicated than I thought. I somehow expected that `JSON.generate` would ignore ALL `to_json`, not only those generated by ActiveSupport. Such a complex logic is atypical, it would be interesting to know where it is defined (ActiveSupport or JSON?) and what’s the reason for it.

Today I will try the patched branch, thanks so much!

**Comment by lucaong** [comment]
Could it be because of this? https://github.com/rails/rails/pull/12862

**Comment by lucaong** [comment]
Uhm, it seems that the JSON gem in `JSON.generate` passes in a `JSON::State` as argument when calling `to_json`, but the `Oj` version of `JSON.generate` does not:

```ruby
class Foo
  def to_json(arg = nil)
    puts("Argument: #{arg.inspect}")
    "{}"
  end
end

f = Foo.new

require 'json'

JSON.generate(f)
# Argument: #<JSON::Ext::Generator::State:0x00007fdd7f128810>
# => "{}"

require 'oj'
Oj.mimic_JSON()

JSON.generate(f)
# Argument: nil
# => "{}"
```

Maybe this line is only called in `JSON.dump` but not in `JSON.generate`? https://github.com/ohler55/oj/blob/b95f3118ca2277a18b975395760f45307cc006a5/ext/oj/mimic_json.c#L246

**Comment by ohler55** [comment]
I'll explore that avenue a bit tonight. Sadly it really is not clear cut.

**Comment by lucaong** [comment]
yeah it's a much deeper rabbit hole than I thought. Thanks for looking into this!

**Comment by ohler55** [comment]
It looks like sometimes a Hash is passed as well according to the JSON gem unit tests. In any case adding a State if neither a Hash or State is already included seems to work. Feel free to try the current generate-special branch.

**Comment by lucaong** [comment]
I confirm that the `generate-special` branch fixes the issue and behaves like the `JSON` gem. It also resolves https://github.com/getsentry/sentry-ruby/issues/1389, which is the original reason that brought me to file this issue.

**Comment by ohler55** [comment]
Great, I'll cut a release tonight.

**Comment by ohler55** [comment]
Released

**Comment by lucaong** [comment]
Awesome, thanks a lot! Closing this issue.
